### ISSUE TITLE
Relations upload_folders does not exist happening after updating Strapi

### ISSUE THREAD

**Issue by jsanta** [strapi/strapi#14038]
### Required System information

- Node.js version:  16.15.1
- NPM version: 8.11.0
- Strapi version: 4.3.2
- Database: Postgres 12.11, 14.2
- Operating system: Ubuntu Linux 20.04, MacOS Catalina

### Describe the bug

Updated Strapi version from 4.2.0 to 4.3.2, just changed this entries in the `package.json` file :
```
"@strapi/plugin-i18n": "4.3.2",
"@strapi/plugin-users-permissions": "4.3.2",
"@strapi/strapi": "4.3.2",
```
Installed the packages using `yarn install`, and rebuilt the Strapi admin panel using `yarn strapi build`
After a while running the server and navigating in the Content Manager started to get this error message:

```
This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
error: select LENGTH(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE("path", $1, $2), $3, $4), $5, $6), $7, $8), $9, $10), $11, $12), $13, $14), $15, $16), $17, $18), $19, $20)) AS depth, COUNT(*) AS occurence from "upload_folders" group by "depth" - relation "upload_folders" does not exist
    at Parser.parseErrorMessage (/opt/queplan/apps/strapi.new/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/opt/queplan/apps/strapi.new/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/opt/queplan/apps/strapi.new/node_modules/pg-protocol/dist/parser.js:39:38)
    at TLSSocket.<anonymous> (/opt/queplan/apps/strapi.new/node_modules/pg-protocol/dist/index.js:11:42)
    at TLSSocket.emit (node:events:527:28)
    at TLSSocket.emit (node:domain:475:12)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at TLSSocket.Readable.push (node:internal/streams/readable:228:10)
    at TLSWrap.onStreamRead (node:internal/stream_base_commons:190:23)
```

Checked the database tables and the upload_folders  table does exist.

### Steps to reproduce the behavior

Sorry about not having the precise steps, it happens after a while when navigating inside the Content Manager.

### Expected behavior

- No errors are expected due to the  database already seem to have all the required relations.
- Double checking the  existing database relations  before  sending an error.
- A more explicit error message in order to be able to look for a workaround.

**Comment by petersg83** [comment]
Thank you for reporting this issue. Can you tell us how many folders and assets you have in your ML please?

**Comment by petersg83** [comment]
Also, do you see any difference in the names? Some uppercase letters that should be lowercase or else?

**Comment by kasonde** [comment]
Hi @jsanta quick que: could you confirm that you had media already uploaded in your application before the update?

**Comment by jsanta** [comment]
Ok some answers.
I did have some media  uploaded on my  original installation, no more than 10  images.
No folders existed (as folders where introduced  after version 4.2).
After placing the suggested code for more info the server won't start at all, showing the exact error I mentioned previously.

**Comment by kasonde** [comment]
Odd, I can't seem to reproduce it all. I'm gonna change the status of the issue for now but I'm curious to know what's going on. Do you mind updating the issue to add in the exact Postgres version and Linux OS version and info you're using? It could be helpful for us.

**Comment by jsanta** [comment]
Don't think the Postgres version or the Linux version are so much relevant here, as the same happens on my Mac (Catalina) with Postgres 14.2, where I am developing.
But just in case, Strapi is deployed on an Ubuntu Linux 20.04.03 LTS, connected to a Postgres 12.11 database.
Same yarn.lock and database schema are being used on both development and deployment installations.

**Comment by smoothdvd** [comment]
I met this issue also. Upgrade from 4.2.3 to 4.3.2.

```
[2022-08-10 09:38:04.265] error: insert into "files" ("alternative_text", "caption", "created_at", "created_by_id", "ext", "folder_path", "formats", "hash", "height", "mime", "name", "provider", "size", "updated_at", "updated_by_id", "url", "width") values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17) returning "id" - column "folder_path" of relation "files" does not exist
error: insert into "files" ("alternative_text", "caption", "created_at", "created_by_id", "ext", "folder_path", "formats", "hash", "height", "mime", "name", "provider", "size", "updated_at", "updated_by_id", "url", "width") values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17) returning "id" - column "folder_path" of relation "files" does not exist
    at Parser.parseErrorMessage (/Users/alexgao/Dev/hangzhou-calendar-strapi/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/Users/alexgao/Dev/hangzhou-calendar-strapi/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/Users/alexgao/Dev/hangzhou-calendar-strapi/node_modules/pg-protocol/dist/parser.js:39:38)
    at Socket.<anonymous> (/Users/alexgao/Dev/hangzhou-calendar-strapi/node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:527:28)
    at Socket.emit (node:domain:475:12)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)
```

And there is no upload_folders, upload_folders_parent_links, files_folder_links tables created, and file table has no folder_path column added.

**Comment by petersg83** [comment]
Are you able to correctly use the folder feature ? Create folders, put files in them etc.?

**Comment by petersg83** [comment]
Just to gather a maximum amount of info : are you using typescript or javascript?

**Comment by jsanta** [comment]
Javascript.
More information, recently upgraded to 4.3.3 . On my production environment same error (trying on a different port just in case I screw things up). Locally had so many problems upgrading that I decided to make a clean database install and then copy all production records. That worked well, locally I haven't seen the error again, but cannot do that on production (actually I can, but  solving further client headbanging due to the site being down will be a huge issue).

**Comment by jsanta** [comment]
I think the only way of doing a near flawless upgrade is to generate a clean project (using a clean db schema) and afterwards 3 steps:

1. replace package.json to install all plugins
2. copy/restore config, api and components folders from the to-be-upgraded version -> run strapi -> stop strapi
3. populate the database from the old-schema -> run strapi -> pray

or just stick to the last working version you installed...  
Will try locally with Strapi 4.3.4 (just out of the oven), deleting node_modules.

**Comment by ronildo** [comment]
I also had the same issue, one I think I notice that might be the issue.
I have the schema.json customized, here is mine, I have added a column called duration (I need to save the duration of the mp3 files)

```json
{
  "collectionName": "files",
  "info": {
    "singularName": "file",
    "pluralName": "files",
    "displayName": "File",
    "description": ""
  },
  "options": {},
  "pluginOptions": {
    "content-manager": {
      "visible": false
    },
    "content-type-builder": {
      "visible": false
    }
  },
  "attributes": {
    "name": {
      "type": "string",
      "configurable": false,
      "required": true
    },
    "alternativeText": {
      "type": "string",
      "configurable": false
    },
    "caption": {
      "type": "string",
      "configurable": false
    },
    "width": {
      "type": "integer",
      "configurable": false
    },
    "height": {
      "type": "integer",
      "configurable": false
    },
    "duration": {
      "type": "integer",
      "configurable": false
    },
    "formats": {
      "type": "json",
      "configurable": false
    },
    "hash": {
      "type": "string",
      "configurable": false,
      "required": true
    },
    "ext": {
      "type": "string",
      "configurable": false
    },
    "mime": {
      "type": "string",
      "configurable": false,
      "required": true
    },
    "size": {
      "type": "decimal",
      "configurable": false,
      "required": true
    },
    "url": {
      "type": "string",
      "configurable": false,
      "required": true
    },
    "previewUrl": {
      "type": "string",
      "configurable": false
    },
    "provider": {
      "type": "string",
      "configurable": false,
      "required": true
    },
    "provider_metadata": {
      "type": "json",
      "configurable": false
    },
    "related": {
      "type": "relation",
      "relation": "morphToMany",
      "configurable": false
    }
  }
}
```

**Comment by derrickmehaffy** [comment]
> I also have had the same issue, while trying to update to the latest version, one I think I notice that might be the issue.
> I have the schema.json customized, here is mine, I have added a column called duration (I need to save the duration of the mp3 files)
> 
> ```json
> {
>   "collectionName": "files",
>   "info": {
>     "singularName": "file",
>     "pluralName": "files",
>     "displayName": "File",
>     "description": ""
>   },
>   "options": {},
>   "pluginOptions": {
>     "content-manager": {
>       "visible": false
>     },
>     "content-type-builder": {
>       "visible": false
>     }
>   },
>   "attributes": {
>     "name": {
>       "type": "string",
>       "configurable": false,
>       "required": true
>     },
>     "alternativeText": {
>       "type": "string",
>       "configurable": false
>     },
>     "caption": {
>       "type": "string",
>       "configurable": false
>     },
>     "width": {
>       "type": "integer",
>       "configurable": false
>     },
>     "height": {
>       "type": "integer",
>       "configurable": false
>     },
>     "duration": {
>       "type": "integer",
>       "configurable": false
>     },
>     "formats": {
>       "type": "json",
>       "configurable": false
>     },
>     "hash": {
>       "type": "string",
>       "configurable": false,
>       "required": true
>     },
>     "ext": {
>       "type": "string",
>       "configurable": false
>     },
>     "mime": {
>       "type": "string",
>       "configurable": false,
>       "required": true
>     },
>     "size": {
>       "type": "decimal",
>       "configurable": false,
>       "required": true
>     },
>     "url": {
>       "type": "string",
>       "configurable": false,
>       "required": true
>     },
>     "previewUrl": {
>       "type": "string",
>       "configurable": false
>     },
>     "provider": {
>       "type": "string",
>       "configurable": false,
>       "required": true
>     },
>     "provider_metadata": {
>       "type": "json",
>       "configurable": false
>     },
>     "related": {
>       "type": "relation",
>       "relation": "morphToMany",
>       "configurable": false
>     }
>   }
> }
> ```

Indeed a customization like this would conflict with us releasing the folders feature as you likely didnt check on the changes to the file model schema.

**Comment by jsanta** [comment]
Certain plugins do change the schemas (eg url_alias) and I suspect there may be many more.
Perhaps focusing in some kind of schema normalizing tool for these kind of upgrades, 'cause a minor version update should not break things like this.

**Comment by derrickmehaffy** [comment]
> Certain plugins do change the schemas (eg url_alias) and I suspect there may be many more.
> Perhaps focusing in some kind of schema normalizing tool for these kind of upgrades, 'cause a minor version update should not break things like this.

Well this was a feature release, there are programmatic ways for plugins (and users) to add fields to models that would not be breaking but it sounds like that was not done here.

We do not consider something like this to be a breaking change.

**Comment by ronildo** [comment]
Can you point me out where/how should I make a customization like I did?

**Comment by jsanta** [comment]
Is there any way of programatically checking if old version entities are matching correctly any model changes? That would be a one time run when the updating is done.

**Comment by jsanta** [comment]
By the way, the error points out that the upload_folders relation is missing, not the table. I guess there may be some relation between some tables that was not created in the version upgrade. Just can't figure out where the ER schema differences are.
Is there any base schema diagram available somewhere? I think that would be really helpful.

**Comment by leo-diehl** [comment]
I'm having exactly the same issue. Upgraded the version, already had media uploaded,
`Error: ER_NO_SUCH_TABLE: Table 'strapi.upload_folders' doesn't exist`

**Comment by leo-diehl** [comment]
Update about this issue: it happened to me after upgrading strapi from `4.3.4` to `4.3.6`.
I downgraded it to `4.3.4` and the bug isn't present anymore.

**Comment by jsanta** [comment]
Just updated to version 4.3.6 
The error keeps jumping on, the only way I've found to trigger it is navigating in the Media Library. Cannot trace to the line of code where it is happenning in order to  at least place a console.log to know what's going on.

Anybody knows where the:

`select LENGTH(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE("path", $1, $2), $3, $4), $5, $6), $7, $8), $9, $10), $11, $12), $13, $14), $15, $16), $17, $18), $19, $20)) AS depth, COUNT(*) AS occurence from "upload_folders" group by "depth"`

query is built?

**Comment by petersg83** [comment]
Hi @jsanta,
You'll find the corresponding code there: https://github.com/strapi/strapi/blob/main/packages/core/upload/server/services/metrics.js#L52

I haven't had time to work on it yet. Considering what was said, I guess it has something to do with the custom schema. Instead of replacing the schema with [overriding method](https://docs.strapi.io/developer-docs/latest/development/plugins-extension.html#extending-a-plugin-s-content-types), can you try to add your custom attributes using the [extension method](https://docs.strapi.io/developer-docs/latest/development/plugins-extension.html#within-the-extensions-folder) ?

**Comment by ronildo** [comment]
> Hi @jsanta, You'll find the corresponding code there: https://github.com/strapi/strapi/blob/main/packages/core/upload/server/services/metrics.js#L52
> 
> I haven't had time to work on it yet. Considering what was said, I guess it has something to do with the custom schema. Instead of replacing the schema with [overriding method](https://docs.strapi.io/developer-docs/latest/development/plugins-extension.html#extending-a-plugin-s-content-types), can you try to add your custom attributes using the [extension method](https://docs.strapi.io/developer-docs/latest/development/plugins-extension.html#within-the-extensions-folder) ?

Thanks for you reply, I will try to fix my build today.

**Comment by jsanta** [comment]
> Hi @jsanta, You'll find the corresponding code there: https://github.com/strapi/strapi/blob/main/packages/core/upload/server/services/metrics.js#L52
> 
> I haven't had time to work on it yet. Considering what was said, I guess it has something to do with the custom schema. Instead of replacing the schema with [overriding method](https://docs.strapi.io/developer-docs/latest/development/plugins-extension.html#extending-a-plugin-s-content-types), can you try to add your custom attributes using the [extension method](https://docs.strapi.io/developer-docs/latest/development/plugins-extension.html#within-the-extensions-folder) ?

The thing is my entities do not have any custom attributes, besides the ones that may have been added by some plugins I'm using.
Is there any way to manually check admitted attributes? Hopefully programatically, otherwise the only chance is comparing visually against the  generated json.

**Comment by petersg83** [comment]
Yes :)
You can run in your terminal 
```
strapi console
```
and then type 
```
strapi.contentType('plugin::upload.file');
```

**Comment by jsanta** [comment]
Ok further investigation lead to more questions:
1. Why is strapi calculating metrics even is telemetry is disabled?
2. Why are Promises  used in the optimistic way? This case is just the evidence that tells  async await may not  be the best way of managing Promises. I'll try some simple  error handling and some query debugging to see what happening underneath.

Ran the strapi console command, didn't detect anything weird. Relations seem to exist so I'm  chasing my tail right now.

**Comment by petersg83** [comment]
1. They are calculated but not sent, the check of if the telemetry is enabled or not is done later in the code. The best would be to check before doing the calculating I agree.
2. What part of the code are you referencing to? I'm not sure to understand what you refer to

So the schema was not changed by a plugin? It's so weird that everything works except that

**Comment by jsanta** [comment]
File: `./node_modules/@strapi/plugin-upload/server/services/metrics.js`

```
const folderLevelsArray = (
      await knex(folderTable)
        .select(
          knex.raw(
            `LENGTH(${keepOnlySlashesSQLString}) AS depth, COUNT(*) AS occurence`,
            queryParams
          )
        )
        .groupBy('depth')
    ).map((map) => ({ depth: Number(map.depth), occurence: Number(map.occurence) }));
```
got  this on line 51~
and on line 94 ~

```
async sendMetrics() {
    const metrics = await this.computeMetrics();
```

Those await calls are not handling possible errors, ths making it even difficult for  debbuging.

**Comment by petersg83** [comment]
yes indeed there is no `try ... catch`. It would be cleaner to put one in the cron function in `registerCron` and only log the error.

**Comment by derrickmehaffy** [comment]
Heads up to @petersg83 an EE customer with internal TID1864 also ran into this issue. Here is a snipped stack log:

```
error: select LENGTH(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE("path", $1, $2), $3, $4), $5, $6), $7, $8), $9, $10), $11, $12), $13, $14), $15, $16), $17, $18), $19, $20)) AS depth, COUNT(*) AS occurence from "upload_folders" group by "depth" - relation "upload_folders" does not exist
 at Parser.parseErrorMessage (-SNIP-/node_modules/pg-protocol/dist/parser.js:287:98)
 at Parser.handlePacket (-SNIP-/node_modules/pg-protocol/dist/parser.js:126:29)
 at Parser.parse (-SNIP-/node_modules/pg-protocol/dist/parser.js:39:38)
 at TLSSocket.<anonymous> (-SNIP-/node_modules/pg-protocol/dist/index.js:11:42)
 at TLSSocket.emit (events.js:400:28)
 at TLSSocket.emit (domain.js:475:12)
 at addChunk (internal/streams/readable.js:293:12)
 at readableAddChunk (internal/streams/readable.js:267:9)
 at TLSSocket.Readable.push (internal/streams/readable.js:206:10)
 at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)
```

**Comment by jsanta** [comment]
Hardcoded an empty catch() on the computeMetrics await and captured the error, but that is just eye candy as it is still happening.

**Comment by petersg83** [comment]
Hello everyone!
I think I have a lead for a fix.
Do you confirm that you use a Postgres schema different than `public`?

**Comment by jsanta** [comment]
On our case yes, we are using a different schema on our Postgres database.

**Comment by petersg83** [comment]
It should be fixed in 4.5.2 :)
